libs/agno/agno/db/postgres/postgres.py

```diff
@@ -1162,6 +1189,210 @@ def delete_user_memory(self, memory_id: str) -> bool:
 
     # -- Metrics methods --
 
+    def _bulk_upsert_metrics(self, table: Table, metrics_records: list[dict]) -> list[dict]:
+        if not metrics_records:
+            return []
+
+        results = []
+        with self.Session() as sess, sess.begin():
+            stmt = postgresql.insert(table)
+
+            # Columns to update in case of conflict
+            update_columns = {
+                col.name: stmt.excluded[col.name]
+                for col in table.columns
+                if col.name not in ["id", "date", "created_at"]
+            }
+
+            stmt = stmt.on_conflict_do_update(index_elements=["date"], set_=update_columns).returning(table)
+            result = sess.execute(stmt, metrics_records)
+            results = [row._mapping for row in result.fetchall()]
+            sess.commit()
+
+        return results
+
+    def _calculate_date_metrics(self, date_to_process: date, sessions_data: dict) -> dict:
+        """Calculate metrics for the given single date"""
+        metrics = {
+            "users_count": 0,
+            "agent_sessions_count": 0,
+            "team_sessions_count": 0,
+            "workflow_sessions_count": 0,
+            "agent_runs_count": 0,
+            "team_runs_count": 0,
+            "workflow_runs_count": 0,
+        }
+        token_metrics = {
+            "input_tokens": 0,
+            "output_tokens": 0,
+            "total_tokens": 0,
+            "audio_tokens": 0,
+            "input_audio_tokens": 0,
+            "output_audio_tokens": 0,
+            "cached_tokens": 0,
+            "cache_write_tokens": 0,
```

Comment:
```suggestion
            "audio_total_tokens": 0,
            "input_audio_tokens": 0,
            "output_audio_tokens": 0,
            "cached_read_tokens": 0,
            "cache_write_tokens": 0,
```

We need to make this change wider actually, some tech debt. You can decide whether we make it now or later.

Reply:
Gotcha. Will handle it outside of this PR